infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/CorfuLogReplicationRuntime.java

```diff
@@ -161,24 +161,24 @@ public class CorfuLogReplicationRuntime {
      */
     public CorfuLogReplicationRuntime(LogReplicationRuntimeParameters parameters,
                                       LogReplicationMetadataManager metadataManager, LogReplicationUpgradeManager upgradeManager,
-                                      LogReplicationSession session, LogReplicationContext replicationContext) {
+                                      LogReplicationSession session, LogReplicationContext replicationContext,
+                                      LogReplicationClientServerRouter router) {
         this.remoteClusterId = session.getSinkClusterId();
         this.session = session;
-        this.router = new LogReplicationClientRouter(parameters, this);
-        this.router.addClient(new LogReplicationHandler());
-        this.sourceManager = new LogReplicationSourceManager(parameters, new LogReplicationClient(router, session.getSinkClusterId()),
-                metadataManager, upgradeManager, session, replicationContext);
+        this.router = router;
+        this.sourceManager = new LogReplicationSourceManager(parameters,router, metadataManager, upgradeManager,
+                session, replicationContext);
         this.connectedNodes = new HashSet<>();
```

Comment:
if Source and not connection initiator, the connected nodes will be only the leader node, right?